pkg/credentialprovider/aws/aws_credentials.go

```diff
@@ -61,172 +103,188 @@ type ecrTokenGetter struct {
-// This way, we avoid timeouts waiting for a non-existent provider.
-func RegisterCredentialsProvider(region string) {
-	klog.V(4).Infof("registering credentials provider for AWS region %q", region)
+// func expiresAtFunc(obj interface{}) (time.Time, error) {
+// 	expiresAt := obj.(cacheEntry).expiresAt
+// 	return expiresAt, nil
+// }
 
-	credentialprovider.RegisterCredentialProvider("aws-ecr-"+region,
-		&lazyEcrProvider{
-			region:    region,
-			regionURL: registryURL(region),
-		})
+// stringKeyFunc is a string as cache key function
+func stringKeyFunc(obj interface{}) (string, error) {
+	key := obj.(cacheEntry).host
+	return key, nil
 }
 
-// Enabled implements DockerConfigProvider.Enabled for the lazy provider.
-// Since we perform no checks/work of our own and actualProvider is only created
-// later at image pulling time (if ever), always return true.
-func (p *lazyEcrProvider) Enabled() bool {
-	return true
+func (p *ecrExpirationPolicy) IsExpired(entry *cache.TimestampedEntry) bool {
+	expiresAt := entry.Obj.(cacheEntry).expiresAt
+	return expiresAt.Before(time.Now()) //TODO clear before expiration
 }
 
-// LazyProvide implements DockerConfigProvider.LazyProvide. It will be called
-// by the client when attempting to pull an image and it will create the actual
-// provider only when we actually need it the first time.
-func (p *lazyEcrProvider) LazyProvide() *credentialprovider.DockerConfigEntry {
-	if p.actualProvider == nil {
-		klog.V(2).Infof("Creating ecrProvider for %s", p.region)
-		p.actualProvider = &credentialprovider.CachingDockerConfigProvider{
-			Provider: newEcrProvider(p.region, nil),
-			// Refresh credentials a little earlier than expiration time
-			Lifetime: 11*time.Hour + 55*time.Minute,
-		}
-		if !p.actualProvider.Enabled() {
-			return nil
-		}
+/*GET CREDENTIALS*/
+func (p *ecrProvider) getFromCache(parsed *parsedURL) (credentialprovider.DockerConfig, bool) {
+	klog.Infof("Checking cache for credentials for %v", parsed.host)
+	cfg := credentialprovider.DockerConfig{}
+	obj, exists, err := p.cache.GetByKey(parsed.host)
+	if err != nil {
+		klog.Infof("unable to get credentials from cache for %v %v", parsed.host, err)
+		return cfg, exists
 	}
-	entry := p.actualProvider.Provide()[p.regionURL]
-	return &entry
+	if exists {
+		entry := obj.(cacheEntry)
+		klog.Info("Credentials found in cache")
+		fmt.Println("in cache")
+		cfg[entry.host] = entry.credentials
+	} else {
+		fmt.Println("not in cache") //TODO
+	}
+	return cfg, exists
 }
 
-// Provide implements DockerConfigProvider.Provide, creating dummy credentials.
-// Client code will call Provider.LazyProvide() at image pulling time.
-func (p *lazyEcrProvider) Provide() credentialprovider.DockerConfig {
-	entry := credentialprovider.DockerConfigEntry{
-		Provider: p,
-	}
+func (p *ecrProvider) getFromECR(parsed *parsedURL) credentialprovider.DockerConfig {
+	klog.Infof("Getting credentials from ECR for %v", parsed.host)
```

Comment:
same comment as @mcurte above, this is too verbose

Reply:
Do you think there should just be one Infof for getting credentials in provide or none? These could be Debugf?

Reply:
yeah I think Debugf would be fine, or Info with --v=6 (or higher)

Reply:
Why all the way to 6? Looking at other providers theirs were a lot lower--currently set to log v(3). IDK if higher means larger # or lower XD.

Reply:
hmm... I guess it depends on how frequently we are hitting this code path. --v=3 I think is fine for the uncached code path, but I would expect something higher (i.e more verbose) for the cached code path.